internal/cache/cache.go

```diff
@@ -197,61 +197,51 @@ func (c *LoadingCache) Close() {
 	close(c.done)
 }
 
-// keysWithTs includes list of keys with ts. This is for sorting keys
-// in order to provide least recently added sorting for size-based eviction
-type keysWithTs []struct {
-	key string
-	ts  time.Time
+// removeOldest removes the oldest item from the cache. Has to be called with lock!
+func (c *LoadingCache) removeOldest() {
+	ent := c.evictList.Back()
+	if ent != nil {
+		c.removeElement(ent)
+	}
 }
 
-// purge records > maxKeys. Has to be called with lock!
-// call with maxKeys 0 will only clear expired entries, with -1 will clear everything.
-func (c *LoadingCache) purge(maxKeys int64) {
-	kts := keysWithTs{}
-
-	for key, value := range c.data {
-		// ttl eviction
-		if time.Now().After(c.data[key].expiresAt) {
-			delete(c.data, key)
-			if c.onEvicted != nil {
-				c.onEvicted(key, value.data)
-			}
-		}
-
-		// prepare list of keysWithTs for size eviction
-		if maxKeys == allKeys || (maxKeys > 0 && int64(len(c.data)) > maxKeys) {
-			ts := c.data[key].expiresAt // for no-LRU sort by expiration time
-			if c.isLRU {                // for LRU sort by read time
-				ts = c.data[key].lastReadAt
-			}
-
-			kts = append(kts, struct {
-				key string
-				ts  time.Time
-			}{key, ts})
-		}
+// Keys returns a slice of the keys in the cache, from oldest to newest. Has to be called with lock!
+func (c *LoadingCache) keys() []string {
+	keys := make([]string, 0, len(c.data))
+	for ent := c.evictList.Back(); ent != nil; ent = ent.Prev() {
+		keys = append(keys, ent.Value.(*cacheItem).key)
 	}
+	return keys
+}
 
-	// size eviction
-	size := int64(len(c.data))
-	if maxKeys == allKeys { // evict everything
-		maxKeys = 0
+// removeElement is used to remove a given list element from the cache. Has to be called with lock!
+func (c *LoadingCache) removeElement(e *list.Element) {
+	c.evictList.Remove(e)
+	kv := e.Value.(*cacheItem)
+	delete(c.data, kv.key)
+	if c.onEvicted != nil {
+		c.onEvicted(kv.key, kv.value)
 	}
-	if len(kts) > 0 {
-		sort.Slice(kts, func(i int, j int) bool { return kts[i].ts.Before(kts[j].ts) })
-		for d := 0; int64(d) < size-maxKeys; d++ {
-			key := kts[d].key
-			value := c.data[key].data
-			delete(c.data, key)
-			if c.onEvicted != nil {
-				c.onEvicted(key, value)
-			}
+}
+
+// deleteExpired deletes expired records. Has to be called with lock!
+func (c *LoadingCache) deleteExpired() {
```

Comment:
not sure what expired exactly means for LRU cache. TTL (expiresAt) shoudn't play any role in this case

Reply:
It works only if `Expired` is set to something less than `noEvictionTTL`. The function does exactly what is written here: deletes expired entries.
It won't work for the LRU cache without the expiration.

Reply:
the point I'm making - LRU should not be tied to expiration in any way. Maybe it was the part of my cache implementation, but we want to simplify it and make LRU really limited by max size only and evict in LRU mode only based on max keys and the total size (indirectly). Supporting TTL in LRU mode makes all of this way too confusing